doc/userguide/suricatasc.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "SURICATASC" "1" "2025-02-11" "8.0.0-dev" "Suricata"
.SH NAME
suricatasc \- Tool to interact via unix socket
.SH SYNOPSIS
.sp
\fBsuricatasc\fP
.SH DESCRIPTION
.sp
Suricata socket control tool
.SH COMMANDS
.INDENT 0.0
.TP
.B shutdown
Shut Suricata instance down.
.UNINDENT
.INDENT 0.0
.TP
.B command\-list
List available commands.
.UNINDENT
.INDENT 0.0
.TP
.B help
Get help about the available commands.
.UNINDENT
.INDENT 0.0
.TP
.B version
Print the version of Suricata instance.
.UNINDENT
.INDENT 0.0
.TP
.B uptime
Display the uptime of Suricata.
.UNINDENT
.INDENT 0.0
.TP
.B running\-mode
Display running mode. This can either be \fIworkers\fP, \fIautofp\fP or \fIsingle\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B capture\-mode
Display the capture mode. This can be either of \fIPCAP_DEV\fP,
\fIPCAP_FILE\fP, \fIPFRING(DISABLED)\fP, \fINFQ\fP, \fINFLOG\fP, \fIIPFW\fP, \fIERF_FILE\fP,
\fIERF_DAG\fP, \fIAF_PACKET_DEV\fP, \fINETMAP(DISABLED)\fP, \fIUNIX_SOCKET\fP or
\fIWINDIVERT(DISABLED)\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B conf\-get <variable>
Get configuration value for a given variable. Variable to be provided can be
either of the configuration parameters that are written in suricata.yaml.
.UNINDENT
.INDENT 0.0
.TP
.B dump\-counters
Dump Suricata\(aqs performance counters.
.UNINDENT
.INDENT 0.0
.TP
.B ruleset\-reload\-rules
Reload the ruleset and wait for completion.
.UNINDENT
.INDENT 0.0
.TP
.B reload\-rules
Alias .. describe \fIruleset\-reload\-rules\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B ruleset\-reload\-nonblocking
Reload ruleset and proceed without waiting.
.UNINDENT
.INDENT 0.0
.TP
.B ruleset\-reload\-time
Return time of last reload.
.UNINDENT
.INDENT 0.0
.TP
.B ruleset\-stats
Display the number of rules loaded and failed.
.UNINDENT
.INDENT 0.0
.TP
.B ruleset\-failed\-rules
Display the list of failed rules.
.UNINDENT
.INDENT 0.0
.TP
.B register\-tenant\-handler <id> <htype> [hargs]
Register a tenant handler with the specified mapping.
.UNINDENT
.INDENT 0.0
.TP
.B unregister\-tenant\-handler <id> <htype> [hargs]
Unregister a tenant handler with the specified mapping.
.UNINDENT
.INDENT 0.0
.TP
.B register\-tenant <id> <filename>
Register tenant with a particular ID and filename.
.UNINDENT
.INDENT 0.0
.TP
.B reload\-tenant <id> [filename]
Reload a tenant with specified ID. A filename to a tenant yaml can be
specified. If it is omitted, the original yaml that was used to load
/ last reload the tenant is used.
.UNINDENT
.INDENT 0.0
.TP
.B reload\-tenants
Reload all registered tenants by reloading their yaml.
.UNINDENT
.INDENT 0.0
.TP
.B unregister\-tenant <id>
Unregister tenant with a particular ID.
.UNINDENT
.INDENT 0.0
.TP
.B add\-hostbit <ipaddress> <hostbit> <expire>
Add hostbit on a host IP with a particular bit name and time of expiry.
.UNINDENT
.INDENT 0.0
.TP
.B remove\-hostbit <ipaddress> <hostbit>
Remove hostbit on a host IP with specified IP address and bit name.
.UNINDENT
.INDENT 0.0
.TP
.B list\-hostbit <ipaddress>
List hostbit for a particular host IP.
.UNINDENT
.INDENT 0.0
.TP
.B reopen\-log\-files
Reopen log files to be run after external log rotation.
.UNINDENT
.INDENT 0.0
.TP
.B memcap\-set <config> <memcap>
Update memcap value of a specified item.
.UNINDENT
.INDENT 0.0
.TP
.B memcap\-show <config>
Show memcap value of a specified item.
.UNINDENT
.INDENT 0.0
.TP
.B memcap\-list
List all memcap values available.
.UNINDENT
.SH PCAP MODE COMMANDS
.INDENT 0.0
.TP
.B pcap\-file <file> <dir> [tenant] [continuous] [delete\-when\-done]
Add pcap files to Suricata for sequential processing. The generated
log/alert files will be put into the directory specified as second argument.
Make sure to provide absolute path to the files and directory. It is
acceptable to add multiple files without waiting the result.
.UNINDENT
.INDENT 0.0
.TP
.B pcap\-file\-continuous <file> <dir> [tenant] [delete\-when\-done]
Add pcap files to Suricata for sequential processing. Directory will be
monitored for new files being added until there is a use of
\fBpcap\-interrupt\fP or directory is moved or deleted.
.UNINDENT
.INDENT 0.0
.TP
.B pcap\-file\-number
Number of pcap files waiting to get processed.
.UNINDENT
.INDENT 0.0
.TP
.B pcap\-file\-list
List of queued pcap files.
.UNINDENT
.INDENT 0.0
.TP
.B pcap\-last\-processed
Processed time of last file in milliseconds since epoch.
.UNINDENT
.INDENT 0.0
.TP
.B pcap\-interrupt
Terminate the current state by interrupting directory processing.
.UNINDENT
.INDENT 0.0
.TP
.B pcap\-current
Currently processed file.
.UNINDENT
.SH BUGS
.sp
Please visit Suricata\(aqs support page for information about submitting
bugs or feature requests.
.SH NOTES
.INDENT 0.0
.IP \(bu 2
Suricata Home Page
.INDENT 2.0
.INDENT 3.5
\fI\%https://suricata.io/\fP
.UNINDENT
.UNINDENT
.IP \(bu 2
Suricata Support Page
.INDENT 2.0
.INDENT 3.5
\fI\%https://suricata.io/support/\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH COPYRIGHT
2016-2025, OISF
.\" Generated by docutils manpage writer.
.


doc/NEWS
https://suricata.io/news/


python/suricata/sc/suricatasc.py
# Copyright(C) 2012-2023 Open Information Security Foundation

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

try:
    import simplejson as json
except ImportError:
    import json
import readline
import select
import sys
from socket import AF_UNIX, error, socket
from inspect import currentframe

from suricata.sc.specs import argsd

SURICATASC_VERSION = "1.0"
VERSION = "0.2"
INC_SIZE = 1024


def get_linenumber():
    cf = currentframe()
    return cf.f_back.f_lineno


class SuricataException(Exception):
    """
    Generic class for suricatasc exception
    """
    def __init__(self, value):
        super(SuricataException, self).__init__(value)
        self.value = value

    def __str__(self):
        return str(self.value)


class SuricataNetException(SuricataException):
    """
    Exception raised when a network error occurs
    """


class SuricataCommandException(SuricataException):
    """
    Exception raised when the command is incorrect
    """


class SuricataReturnException(SuricataException):
    """
    Exception raised when return message is incorrect
    """


class SuricataCompleter:
    def __init__(self, words):
        self.words = words
        self.generator = None

    def complete(self, text):
        for word in self.words:
            if word.startswith(text):
                yield word

    def __call__(self, text, state):
        if state == 0:
            self.generator = self.complete(text)
        try:
            return next(self.generator)
        except StopIteration:
            return None


class SuricataSC:
    def __init__(self, sck_path, verbose=False):
        self.basic_commands = [
                "shutdown",
                "quit",
                "pcap-file-number",
                "pcap-file-list",
                "pcap-last-processed",
                "pcap-interrupt",
                "iface-list",
                "reload-tenants",
                ]
        self.fn_commands = [
                "pcap-file",
                "pcap-file-continuous",
                "iface-stat",
                "conf-get",
                "unregister-tenant-handler",
                "register-tenant-handler",
                "unregister-tenant",
                "register-tenant",
                "reload-tenant",
                "add-hostbit",
                "remove-hostbit",
                "list-hostbit",
                "memcap-set",
                "memcap-show",
                "dataset-add",
                "dataset-remove",
                "datajson-add",
                "datajson-remove",
                "get-flow-stats-by-id",
                "dataset-clear",
                "dataset-lookup",
                ]
        self.cmd_list = self.basic_commands + self.fn_commands
        self.sck_path = sck_path
        self.verbose = verbose
        self.socket = socket(AF_UNIX)

    def json_recv(self):
        cmdret = None
        data = ""
        while True:
            if sys.version < '3':
                received = self.socket.recv(INC_SIZE)
            else:
                received = self.socket.recv(INC_SIZE).decode('iso-8859-1')

            if not received:
                break

            data += received
            if data.endswith('\n'):
                cmdret = json.loads(data)
                break
        return cmdret

    def send_command(self, command, arguments=None):
        if command not in self.cmd_list and command != 'command-list':
            raise SuricataCommandException("L{}: Command not found: {}".format(get_linenumber(), command))

        cmdmsg = {}
        cmdmsg['command'] = command
        if arguments:
            cmdmsg['arguments'] = arguments
        if self.verbose:
            print("SND: " + json.dumps(cmdmsg))
        cmdmsg_str = json.dumps(cmdmsg) + "\n"
        if sys.version < '3':
            self.socket.send(cmdmsg_str)
        else:
            self.socket.send(bytes(cmdmsg_str, 'iso-8859-1'))

        ready = select.select([self.socket], [], [], 600)
        if ready[0]:
            cmdret = self.json_recv()
        else:
            cmdret = None
        if not cmdret:
            raise SuricataReturnException("L{}: Unable to get message from server".format(get_linenumber))

        if self.verbose:
            print("RCV: "+ json.dumps(cmdret))

        return cmdret

    def connect(self):
        try:
            if self.socket is None:
                self.socket = socket(AF_UNIX)
            self.socket.connect(self.sck_path)
        except error as err:
            raise SuricataNetException("L{}: {}".format(get_linenumber(), err))

        self.socket.settimeout(10)
        #send version
        if self.verbose:
            print("SND: " + json.dumps({"version": VERSION}))
        if sys.version < '3':
            self.socket.send(json.dumps({"version": VERSION}))
        else:
            self.socket.send(bytes(json.dumps({"version": VERSION}), 'iso-8859-1'))

        ready = select.select([self.socket], [], [], 600)
        if ready[0]:
            cmdret = self.json_recv()
        else:
            cmdret = None

        if not cmdret:
            raise SuricataReturnException("L{}: Unable to get message from server".format(get_linenumber()))

        if self.verbose:
            print("RCV: "+ json.dumps(cmdret))

        if cmdret["return"] == "NOK":
            raise SuricataReturnException("L{}: Error: {}".format(get_linenumber(), cmdret["message"]))

        cmdret = self.send_command("command-list")

        # we silently ignore NOK as this means server is old
        if cmdret["return"] == "OK":
            self.cmd_list = cmdret["message"]["commands"]
            self.cmd_list.append("quit")

    def close(self):
        self.socket.close()
        self.socket = None

    def execute(self, command):
        full_cmd = command.split()
        cmd = full_cmd[0]
        cmd_specs = argsd[cmd]
        required_args_count = len([d["required"] for d in cmd_specs if d["required"] and not "val" in d])
        arguments = dict()
        # if all arguments are required in the command then we split at the count
        # this way we can handle last argument containing space (datajson-add for example)
        non_req_args_count = len([d for d in cmd_specs if not d["required"] or "val" in d])
        if non_req_args_count == 0:
            full_cmd = command.split(maxsplit=required_args_count)
        for c, spec in enumerate(cmd_specs, 1):
            spec_type = str if "type" not in spec else spec["type"]
            if spec["required"]:
                if spec.get("val"):
                    arguments[spec["name"]] = spec_type(spec["val"])
                    continue
                try:
                    arguments[spec["name"]] = spec_type(full_cmd[c])
                except IndexError:
                    phrase = " at least" if required_args_count != len(cmd_specs) else ""
                    msg = "Missing arguments: expected{} {}".format(phrase, required_args_count)
                    raise SuricataCommandException("L{}: {}".format(get_linenumber(), msg))
                except ValueError as ve:
                    raise SuricataCommandException("L{}: Erroneous arguments: {}".format(get_linenumber(), ve))
            elif c < len(full_cmd):
                arguments[spec["name"]] = spec_type(full_cmd[c])
        return cmd, arguments

    def parse_command(self, command):
        arguments = None
        cmd = command.split()[0] if command else None
        if cmd in self.cmd_list:
            if cmd in self.fn_commands:
                cmd, arguments = getattr(self, "execute")(command=command)
        else:
            raise SuricataCommandException("L{}: Unknown command: {}".format(get_linenumber(), command))
        return cmd, arguments

    def interactive(self):
        print("Command list: " + ", ".join(self.cmd_list))
        try:
            readline.set_completer(SuricataCompleter(self.cmd_list))
            readline.set_completer_delims(";")
            readline.parse_and_bind('tab: complete')
            while True:
                if sys.version < '3':
                    command = raw_input(">>> ").strip()
                else:
                    command = input(">>> ").strip()
                if command == "quit":
                    break
                if len(command.strip()) == 0:
                    continue
                try:
                    cmd, arguments = self.parse_command(command)
                except SuricataCommandException as err:
                    print(err)
                    continue
                try:
                    cmdret = self.send_command(cmd, arguments)
                except IOError as err:
                    # try to reconnect and resend command
                    print("Connection lost, trying to reconnect")
                    try:
                        self.close()
                        self.connect()
                    except (SuricataNetException, SuricataReturnException) as err:
                        print(err.value)
                        continue
                    cmdret = self.send_command(cmd, arguments)
                except (SuricataCommandException, SuricataReturnException) as err:
                    print("An exception occured: " + str(err.value))
                    continue
                #decode json message
                if cmdret["return"] == "NOK":
                    print("Error:")
                    print(json.dumps(cmdret["message"], sort_keys=True, indent=4, separators=(',', ': ')))
                else:
                    print("Success:")
                    print(json.dumps(cmdret["message"], sort_keys=True, indent=4, separators=(',', ': ')))
        except KeyboardInterrupt:
            print("[!] Interrupted")
            sys.exit(0)
